pkg/compactor/blocks_cleaner.go

```diff
@@ -129,54 +137,175 @@ func NewBlocksCleaner(
 			Name: "cortex_bucket_index_last_successful_update_timestamp_seconds",
 			Help: "Timestamp of the last successful update of a tenant's bucket index.",
 		}, []string{"user"}),
+		tenantBlocksCleanedTotal: promauto.With(reg).NewCounterVec(prometheus.CounterOpts{
+			Name: "cortex_bucket_blocks_cleaned_total",
+			Help: "Total number of blocks deleted for a tenant.",
+		}, commonLabels),
+		tenantCleanDuration: promauto.With(reg).NewGaugeVec(prometheus.GaugeOpts{
+			Name: "cortex_bucket_clean_duration_seconds",
+			Help: "Duration of cleaner runtime for a tenant in seconds",
+		}, commonLabels),
 	}
 
-	c.Service = services.NewTimerService(cfg.CleanupInterval, c.starting, c.ticker, nil)
+	c.Service = services.NewBasicService(c.starting, c.loop, nil)
 
 	return c
 }
 
+type cleanerJob struct {
+	users     []string
+	timestamp int64
+}
+
+func (c *BlocksCleaner) SetRingLifecyclerID(ringLifecyclerID string) {
+	c.ringLifecyclerID = ringLifecyclerID
+}
+
 func (c *BlocksCleaner) starting(ctx context.Context) error {
 	// Run a cleanup so that any other service depending on this service
 	// is guaranteed to start once the initial cleanup has been done.
-	c.runCleanup(ctx, true)
+	activeUsers, deletedUsers, err := c.scanUsers(ctx)
 
+	if err != nil {
+		level.Error(c.logger).Log("msg", "failed to scan users on startup", "err", err.Error())
+		c.runsFailed.WithLabelValues(deletedStatus).Inc()
+		c.runsFailed.WithLabelValues(activeStatus).Inc()
+		return nil
```

Comment:
The reason why we don't return error is for cleaner to retry at `loop`?
This looks weird. I don't know if other Cortex components do the same thing

Reply:
It is not intended to return `nil`. Let me revisit the code and update it.

Reply:
Based on current logic (https://github.com/cortexproject/cortex/blob/master/pkg/compactor/blocks_cleaner.go#L144), cleaner never returns error. If it failed, it would just wait for next tick and run it again.

Reply:
This behavior sounds weird tbh. But let's leave it as it is. I guess if there is any error connecting to object storage like credential issues, it expects compactor to fail from compaction, not from cleaner.